* Useful operations
**  Install from Github
#+BEGIN_SRC R :session *R* :eval yes
library(devtools)
devtools::install_github("hilldr/SeqRetriever/SeqRetriever")
#+END_SRC

#+RESULTS:
: TRUE

** Process package documentation
#+BEGIN_SRC R :session *R* :eval yes
wd <- getwd()
if (wd != "/home/david/development/SeqRetriever/SeqRetriever"){
  setwd("/home/david/development/SeqRetriever/SeqRetriever")
} else {
  print("Already in the SeqRetriever working directory")
}
#file.remove("NAMESPACE")
library(roxygen2)
library(devtools)
document()
#+END_SRC

#+RESULTS:

** Install latest version from local source
#+BEGIN_SRC R :session *R* :eval yes
library(devtools)
setwd("..")
install("SeqRetriever")
#+END_SRC

#+RESULTS:
: TRUE

** Uninstall SeqRetriever
#+BEGIN_SRC R :session *R* :eval yes
remove.packages("SeqRetriever")
#+END_SRC

#+RESULTS:

** Convert Org-mode to github markdown
#+begin_src sh
pandoc --from=org --to=markdown -o README.md SeqRetriever.org 
#+end_src

#+RESULTS:

#+begin_src ditaa :file ./img/workflow.png :cmdline -E
+-------------+   +------------------+   +---------------------------+
+ FASTQ files +-->+ Tophat/Cufflinks +-->+ Cuffnorm output directory +--\
+-------------+   +------------------+   +---------------------------+  |
/-----------------------------------------------------------------------/
|
V
+---------------------------------------+
+ Format R dataframe with SeqDataframe()|
+-----------------+---------------------+
                  |
                  V
+-----------------+--------------------------+
| Retrieve genes of interest with SeqGenes() |
+-----------------+--------------------------+
                  |
                  V
+-----------------+-----------------------------+
| Subset genes bases on statistical significance|
: with SeqStatSubset()                          :
+-----------------+-----------------------------+
                  |
                  |   +----------------------------------+
                  +-->+ Output Boxplots with SeqBoxplot()|
                  |   +----------------------------------+
                  |   +---------------------------------+
                  +-->+ Output Heatmap with SeqHeatmap()|
                  |   +---------------------------------+
                  |   +--------------------------------------------+
                  +-->+ Output CSV with P values with SeqStudents()|
                      +--------------------------------------------+
#+end_src

#+RESULTS:
[[file:./img/workflow.png]]

** Benchmarking SeqDataframe
#+begin_src R :session *R* :results text :exports none :eval yes
library(profvis)
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqDataframe.R")
test <- profvis({    
   slo <- SeqDataframe(dir = "/home/david/Data/RNAseq/HIO_RNAseq/HIO_Ecoli_ES/RESULTS/norm_out")
})
#+end_src

#+RESULTS:

11200 ms
Replace plyr with dplyr
970 ms
Replace read.table with readr::read_delim
380 ms


** Validate results
#+begin_src R :session *R* :exports none :results graphics :file figure.png :width 1000 :height 1000 :eval yes
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqDataframe.R")
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqBoxplot.R")
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqGenes.R")
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqStatSubset.R")
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqStudents.R")
source("/home/david/development/SeqRetriever/SeqRetriever/R/SeqHeatmap.R")

plot <- SeqBoxplot(SeqStatSubset(SeqGenes(gene.names = c("DUOX1","DUOX2","DUOXA2","REG3G","SAA1","NOS2","IL22","CCL20","CXCL1"),
                                          df = SeqDataframe(dir = "/home/david/Data/RNAseq/HIO_RNAseq/HIO_Ecoli_ES/RESULTS/HIO_norm_out")),
                                 group1 = "Ctl", group2 = "Ecoli"))
print(plot)
#+end_src

#+RESULTS:
[[file:figure.png]]

